### DISCUSSION TITLE
Workflow runs not starting after previous workflow completes

### DISCUSSION THREAD

**Discussion by andrewconnell** [community/community#26062] [Actions]
<p>After following <a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#workflow_run" rel="nofollow noopener">the docs</a>, I can’t figure out why my workflow isn’t starting. I have one workflow <code>name: Production build, test, &amp; deploy</code> completing.</p>
<p>But two other workflows setup to run after that workflow completes aren’t starting. They aren’t even showing in the list of workflows on the <strong>Actions</strong> page.</p>
<p>Here’s what one of the child workflows looks like that I’d like to start after the one listed above completes:</p>
<pre><code class="lang-auto">name: Production CDN purge

defaults:
  run:
    shell: bash

on:
  workflow_run:
    workflows: ["Production build, test, &amp; deploy"]
    types:
      - completed
</code></pre>
<p>Does the “parent” workflow have to do anything special for these others to be seen?</p>

**Comment by Simran-B** [comment]
<aside class="quote no-group" data-username="andrewconnell" data-post="1" data-topic="128345">
<div class="title">
<div class="quote-controls"></div>
 andrewconnell:</div>
<blockquote>
<p>They aren’t even showing in the list of workflows on the <strong>Actions</strong> page.</p>
</blockquote>
</aside>
<p>That sounds suspicious. Are they in the right folder (<code>.github/workflows/</code>), on the right branch (default branch, usually <code>master</code>), with the correct file extension (<code>.yml</code> or <code>yaml</code>)? Any special characters or whitespace in the file names?</p>

**Comment by andrewconnell** [comment]
<p>They are in the correct folder and have the correct extension, but they aren’t in the <strong>master</strong> branch. The two files in the red box are the ones that aren’t running when the one with the red arrow comples:</p>
<p><img src="https://user-images.githubusercontent.com/2068657/181095034-811bab88-8e69-469f-bda2-93a54d438fca.png" alt="image" data-base62-sha1="djOoNLZuHkNOHlnp3wQCBXPcbnP" width="552" height="397"></p>
<p>We’re currently working on a branch and testing everything before moving to <strong>master</strong> which will roll things into production. Not being on <strong>master</strong> branch… if that’s our issue then we’ll hold off and try again when we push to production.</p>
<p>For now, we’ve just copied the environment variables &amp; contents of the two “child” workflows into the “parent” workflow as new jobs so things will work.</p>

**Comment by Simran-B** [comment]
<p>I stand corrected, the workflow file has to be in the default branch in order to trigger builds manually:</p>
<blockquote>
<p><a href="https://docs.github.com/en/actions/configuring-and-managing-workflows/configuring-a-workflow#manually-running-a-workflow" rel="nofollow noopener">Manually running a workflow</a><br>
[…]<br>
To trigger the  <code>workflow_dispatch</code>  event on GitHub, your workflow must be in the default branch.</p>
</blockquote>
<p>Regarding automated builds:</p>
<blockquote>
<p><a href="https://docs.github.com/en/actions/configuring-and-managing-workflows/configuring-a-workflow#creating-a-workflow-file" rel="nofollow noopener">Creating a workflow file</a><br>
[…]<br>
4. Commit your changes in the workflow file to the branch where you want your workflow to run.</p>
</blockquote>
<blockquote>
<p><a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#about-workflow-events" rel="nofollow noopener">About workflow events</a><br>
[…]<br>
For example, if the event occurred on a particular repository branch, then the workflow files must be present in the repository on that branch.</p>
</blockquote>
<p>Anyway, the <em>Actions</em> tab should show the workflows from all branches (there’s a branch filter option).</p>
<p>Can you share the <code>on:</code> section of the “Production build, test, &amp; deploy” workflow?</p>

**Comment by andrewconnell** [comment]
<p>So it should run from non <strong>master</strong> branches… that’s what I thought. Here’s what the <code>on:</code> section has for the “Production build, test, &amp; deploy” workflow:</p>
<pre><code class="lang-auto">on:
  push:
    branches:
      - v3-migration-to-hugo
  # schedule: TODO: !GOLIVE!  UNCOMMENT THIS
  #   - cron: '1 12,14,17,21 * * 1,2,3,4,5' TODO: !GOLIVE!  UNCOMMENT THIS
</code></pre>
<p>FWIW, we’re looking to merge everything into <strong>master</strong> tomorrow as our workaround has been sufficient so far.</p>

**Comment by Simran-B** [comment]
<p>Alright. Would still like to know why it’s not working with multiple workflows…</p>
<p>You could try to rename the workflow file to something that was never used before just to be sure, as there might be a bug (it’s about <code>on: workflow_dispatch</code> however):</p>
<aside class="quote" data-post="48" data-topic="121740">
  <div class="title">
    <div class="quote-controls"></div>
    
    <a href="https://github.community/t/cant-trigger-workflow-manually/121740/48">Can't trigger workflow manually</a> <a class="badge-wrapper  bullet" href="/c/code-to-cloud/github-actions/41"><span class="badge-category-parent-bg" style="background-color: #6a737d;"></span><span class="badge-category-bg" style="background-color: #2188ff;"></span><span style="" data-drop-close="true" class="badge-category clear-badge" title="Discussion and questions about GitHub Actions—automation for all of your development workflows, now with world-class CI/CD.">GitHub Actions</span></a>
  </div>
  <blockquote>
    In case it helps someone, it worked after changed the file name! It seems that it is a bug with previous file names with the same name.
  </blockquote>
</aside>

<p>And as a last wild idea, maybe rename the parent workflow to something without ampersand and commas. Also change the <code>workflows</code> array accordingly of course. It’s a string in double quotes in both places, so this should be perfectly fine, but who knows.</p>

**Comment by andrewconnell** [comment]
<p>Sure… and happy to help figure it out. I’ll give these things a shot in the morning… thanks for following up!</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/andrewconnell">@andrewconnell</a>,</p>
<p>There are few points maybe can explain the problems:</p>
<ol>
<li>
<p>The workflow runs on <strong>workflow_run</strong> event must exist on the <strong>master</strong> or default branch. And this workflow will only run on the <strong>master</strong> or default branch, regardless of whether it is triggered by a workflow run on other branches or not.</p>
</li>
<li>
<p>On the <strong>Actions</strong> page of a repository, under the “<strong>All workflows</strong>”, generally the following workflows are listed:</p>
<ul>
<li>
<p>The workflow files exist on any branches, and the workflows have been triggered at least once.</p>
</li>
<li>
<p>The workflow files exist on the <strong>master</strong> or default branch, even if the workflows are never triggered.</p>
</li>
<li>
<p>For some events, especially the <strong>workflow_run</strong> event, if the workflow file does not exist on the <strong>master</strong> or default branch, this workflow will not be listed under the “<strong>All workflows</strong>”, even if the workflow file exists on other branches. This workflow also will be never triggered.</p>
</li>
</ul>
</li>
</ol>

**Reply by basickarl** [reply]
If only this info was more readily available.

**Reply by ahgraber** [reply]
Is this still the case?  It is ~hard~ impossible to test workflows in development branches if this is true. It would be great to have some flag to specify which branch the event-triggered workflow should use (like perhaps the branch that initiated the triggering workflow?)

**Reply by jbnjohnathan** [reply]
I am also interested in this, seems strange that this cannot be tested before merge to main

**Reply by jefffroman** [reply]
+1 , we need to be able to test new workflow dependencies before merging to trunk.

**Reply by arikkfir** [reply]
Indeed - it's hard to test this functionality when it only runs such workflows from `main`...

**Reply by paololazzari** [reply]
I fail to see the usefulness of it given the incredibly silly limitation.

**Comment by andrewconnell** [comment]
<p>I renamed the primary workflow to have only alpha+spaces… but the others still aren’t picking it up:</p>
<p><img src="https://user-images.githubusercontent.com/2068657/181095040-593ed550-88e1-41bb-b69c-1ef7cc2b7cdb.png" alt="image" data-base62-sha1="tK6Cx9fPjpfwKlDfjpXYwlV2orF" width="319" height="236"></p>
<p>But as you said, the workflow runs on the <strong>workflow_run</strong> event must be on <strong>master</strong> (default) branch. Everything is still in a different branch, so it now makes sense they they aren’t working. Once we merge into <strong>master</strong> we’ll give it another shot.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/andrewconnell">@andrewconnell</a>,</p>
<p>How are things going?<br>
If you have any updates or questions about this topic, feel free to tell us.</p>

**Comment by andrewconnell** [comment]
<p>My bad for not following up… we merged to <strong>master</strong> and launched our app and have spent the time since then dealing with the post-launch fires :slight_smile:</p>
<p>Seems once we made it to the <strong>master</strong> (default) branch, everything started working. I wasn’t clear on the docs with that point so that was more of my error I think. Thanks for the help!</p>
